core/src/main/scala/org/apache/spark/rdd/RDD.scala

```diff
@@ -1253,9 +1253,9 @@ abstract class RDD[T: ClassTag](
 
   /**
    * @return true if and only if the RDD contains no elements at all. Note that an RDD
-   *         may be empty even when it has at least 1 partition.
+   *         may be empty even when it has 1 or more partitions.
    */
-  def isEmpty(): Boolean = partitions.length == 0 || take(1).length == 0
+  def isEmpty(): Boolean = partitions.length == 0 || mapPartitions(it => Iterator(!it.hasNext)).reduce(_&&_)
```

Comment:
You might have a point, that this is actually a bug in `take()`.

`sc.parallelize(Seq(1,2,3),1).take(1337)` works fine but and returns Array[Int] = Array(1, 2, 3)`

`sc.parallelize(Seq(),1).take(1)` fails on the other hand.